src/Payer.sol

```diff
@@ -0,0 +1,222 @@
+            // If there's enough balance to pay in full: send the entire amount
+            paymentToken.safeTransfer(account, amount);
+        } else if (availableBalance > 0) {
+            // If there's some balance, but not enough to pay in full
+
+            // 1. Pay what we can
+            paymentToken.safeTransfer(account, availableBalance);
+
+            // 2. Add a debt entry for the remaining amount
+            registerDebt(account, amount - availableBalance);
+        } else {
+            // Zero balance, add a debt entry for the entire amount
+            registerDebt(account, amount);
+        }
+    }
+
+    /// @notice Withdraws the entire balance of `paymentToken` to the owner
+    /// @dev Only the owner is allowed to call this function
+    function withdrawPaymentToken() external onlyOwner {
+        address to = owner();
+        uint256 amount = paymentToken.balanceOf(address(this));
+        paymentToken.safeTransfer(to, amount);
+
+        emit TokensWithdrawn(to, amount);
+    }
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      EXTERNAL FUNCTIONS (ANYONE CAN CALL)
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /// @notice Pays back debt up to `amount` of `paymentToken`. Debt is paid in FIFO order.
+    /// @param amount The maximum amount of tokens to send. This is expected to be the token balance of this contract
+    /// @dev Assumes new entries are add to the 'back' of the queue, meaning oldest entries are at the 'front' of the queue
+    function payBackDebt(uint256 amount) external {
+        // While there are tokens left, and debt entries exist
+        while (amount > 0 && !queue.empty()) {
+            // Get the oldest entry
+            DebtQueue.DebtEntry storage debt = queue.front();
+
+            // Cache storage values
+            uint256 _debtAmount = debt.amount;
+            address _debtAccount = debt.account;
+
+            if (amount < _debtAmount) {
+                // Not enough to cover entire debt, pay what you can and leave
+                uint256 remainingDebt = debt.amount - amount;
+
+                // Update remaining debt in queue
+                debt.amount = remainingDebt;
+
+                // Update total debt
+                totalDebt -= amount;
+
+                // Pay user what we can
+                paymentToken.safeTransfer(_debtAccount, amount);
+                emit PaidBackDebt(_debtAccount, amount, remainingDebt);
+
+                // No more tokens, leave
+                return;
+            } else {
+                // Enough to cover entire debt entry, pay in full and remove from queue
+
+                // Update amount of tokens left to pay back debt
+                amount -= _debtAmount;
+
+                // Update total debt
+                totalDebt -= _debtAmount;
+
+                // Pay user the entire amount
+                paymentToken.safeTransfer(_debtAccount, _debtAmount);
```

Comment:
what if this transfer fails?
this could happen for example if `_debtAccount` is blacklisted by USDC.
this would brick this function.

in this case we would need to withdraw the funds and redeploy and adding the debt entries again. possible but not so fun.

on the other hand, not sure it's worth it to handle cases where transfer fails.

happy to get opinions on this

Reply:
Considered, but ultimately felt low-risk that the DAO would pass a proposal that pays out to a blacklisted address, or that an address would be blacklisted in the time between proposal success and the payout.

Reply:
I could not think of any other scenario in which this transfer could fail. Curious if you came up with any. There's no way that the payee can 'block' the transfer.

Reply:
I could only think of the blacklist scenario in the case of USDC